### ISSUE TITLE
Is this a bug in DistributedExecutionService.CallableProcessor.run()?

### ISSUE THREAD

**Issue by juanavelez** [hazelcast/hazelcast#6283]
I apologize in advance if this is not really a bug but from what I can understand from the code in DistributedExecutionService.CallableProcessor.run() when checking for isCancelled() if true, a null **might** be sent as response which will override the sending of an exception for example in DistributorExecutionService.cancel().

Should the sendResponse in the run()'s finally block be part of the block of code called only it is not cancelled?

```
                sendResponse(result);
                if (!isCancelled()) {
                    finishExecution(name, Clock.currentTimeMillis() - start);
                }
```
Please feel free to change the summary/description for this issue if it is correctly identified as a bug.

Thanks!

**Comment by metanet** [comment]
Hi @juanavelez 

In ``sendResponse``, a flag is atomically set when response is sent. So there won't be a second response.

**Comment by juanavelez** [comment]
@metanet Thanks for your explanation. I agree that a single response will only go back to the caller. My point is that such response in case of a "cancel" event is likely going to be a null instead of the exception that caused the "cancel". 

```
    public boolean cancel(String uuid, boolean interrupt) {
        CallableProcessor processor = submittedTasks.remove(uuid);
        if (processor != null && processor.cancel(interrupt)) {
            processor.sendResponse(new CancellationException());
            getLocalExecutorStats(processor.name).cancelExecution();
            return true;
        }
        return false;
    }
``` 
In this case, even though the ```processor.cancel(interrupt)``` indeed canceled the execution, the call to ```processor.sendResponse(new CancellationException())`` could end up being a noop as it is very likely (as you pointed out) that

```
            } finally {
                if (uuid != null) {
                    submittedTasks.remove(uuid);
                }
                sendResponse(result);
```
occurred first and therefore the "result" being sent back to the caller is a "null" instead of the CancellingException.

**Comment by metanet** [comment]
let me check it again

**Comment by metanet** [comment]
Hi @juanavelez 

You are right about your suspicion. We fixed the problem with #6962 

Thank you for the report
